tool/src/dart/mod.rs

```diff
@@ -438,30 +447,56 @@ impl<'a, 'cx> TyGenContext<'a, 'cx> {
-                needs_temp_arena = true;
-                Some("temp.arena".to_string())
-            } else if let hir::Type::Slice(s) = param.ty {
-                match param_borrow_kind {
-                    ParamBorrowInfo::BorrowedSlice => {
-                        // Slices borrowed in the return value use a custom arena
-                        arenas.push(format!("final {param_name}Arena = _FinalizedArena();").into());
-                        Some(format!("{param_name}Arena.arena"))
-                    }
-                    // Owned slices use the Rust allocator
-                    ParamBorrowInfo::TemporarySlice if s.lifetime().is_none() => None,
-                    ParamBorrowInfo::TemporarySlice => {
-                        // Everyone else uses the temporary arena that keeps stuff alive until the method is called
-                        needs_temp_arena = true;
-                        Some("temp.arena".into())
+            /// Get the name/initializer of the allocator needed for a particular type
+            fn alloc_name(
+                param_ty: &hir::Type<hir::InputOnly>,
+                param_name: &str,
+                param_borrow_kind: &ParamBorrowInfo,
+                needs_temp_arena: &mut bool,
+                arenas: &mut Vec<Cow<str>>,
+            ) -> Option<String> {
+                if let hir::Type::Struct(..) = param_ty {
+                    *needs_temp_arena = true;
+                    Some("temp.arena".to_string())
+                } else if let hir::Type::Slice(s) = param_ty {
+                    match param_borrow_kind {
+                        ParamBorrowInfo::BorrowedSlice => {
+                            // Slices borrowed in the return value use a custom arena
+                            arenas.push(
+                                format!("final {param_name}Arena = _FinalizedArena();").into(),
+                            );
+                            Some(format!("{param_name}Arena.arena"))
+                        }
+                        // Owned slices use the Rust allocator
+                        ParamBorrowInfo::TemporarySlice if s.lifetime().is_none() => None,
+                        ParamBorrowInfo::TemporarySlice => {
+                            // Everyone else uses the temporary arena that keeps stuff alive until the method is called
+                            *needs_temp_arena = true;
+                            Some("temp.arena".into())
+                        }
+                        _ => unreachable!(
+                            "Slices must produce slice ParamBorrowInfo, found {param_borrow_kind:?}"
+                        ),
                     }
-                    _ => unreachable!(
-                        "Slices must produce slice ParamBorrowInfo, found {param_borrow_kind:?}"
-                    ),
+                } else if let hir::Type::DiplomatOption(ref inner) = param_ty {
+                    alloc_name(
```

Comment:
not a fan of the recursion. Can you instead match on `DiplomatOption(Struct)` in addition to `Struct`?

Reply:
This would subtly break if we ever added option-slice support, though?

Reply:
You can match on `DiplomatOption`, and the let-bind the struct with an unreachable else case.

Reply:
Or have a `DiplomatOption(Struct)` case with the `Struct` case, and a `DiplomatOption(..)` case that is unreachable.

Reply:
note that `box` patterns are unstable so you cannot use the same match arm. I cannot do `Struct(s) | DiplomatOption(Struct(s))`, so either way I need _some_ recursion or at least a more scope function for structs.

Reply:
ugh

Reply:
yeah, I started writing it this way first and quickly remembered box patterns were unstable.